tools/tutorial-opt.cpp

```diff
@@ -46,6 +46,12 @@ void polyToLLVMPipelineBuilder(mlir::OpPassManager &manager) {
   manager.addPass(mlir::createConvertFuncToLLVMPass());
   manager.addPass(mlir::createFinalizeMemRefToLLVMConversionPass());
   manager.addPass(mlir::createReconcileUnrealizedCastsPass());
+
+  // Cleanup
+  manager.addPass(mlir::createCanonicalizerPass());
```

Comment:
Is there a particular reason for adding `mlir::createCanonicalizerPass()` two times? There's one on the top which I think was for the passes we had created and there's this one which to my understanding should be dealing with the canonicalization of the other upstream passes we added to this pipeline.

I tried to remove the first one and just keep this last one here and the binary still works correctly (just that there is some difference in the instructions being generated in the end).

Reply:
In general canonicalization must not change the behavior of the program (or else it would be buggy). It will only make things more efficient. Canonicalization patterns apply to particular ops in particular dialects, and higher-level canonicalization patterns tend to remove more unnecessary work than applying lower-level canonicalization patterns, since they capture a larger breadth of the overall program. Running canonicalization after every pass would be even better. But the reason we don't is because we expect it to do nothing, and it adds to the total runtime.